Refactor/introduce change detector

jjelschen opened the pull request:
(builds on changes included in #29, so that pull request should be merged first.)

Introduces a set of `ChangeDetector` implementations, that can be chained. For the test editor, this would be wired as follows:
1. The `GitBasedChangeDetector` runs first. It basically takes on the responsibilities of class `IndexUpdater`, however, I later decided to factor out all language-relevance checks, significantly simplifying the new class.
2. The `GradleBasedChangeDetector` takes over next. It contains (part of) the logic of `GradleProjectIndexUpdater`, and otherwise relies on logic extracted from the `StandaloneBuilder` into `XtextBuilderUtils`. Because of that, it does some filtering out of the box.
3. The `ChangeFilter` terminates the chain. Not strictly a change _detector_ (I'm open for alternative name suggestions for the general interface…), this class is responsible for filtering out anything that should not actually go into the index. It can be wired up with `IndexFilter`s, e.g. `LanguageExtensionBasedIndexFilter` and `SearchPathBasedIndexFilter`.

Commit:
introduce IndexFilter interface

refactoring: move index relevance test from IndexUpdater into a separate class

Commit:
add GitBasedChangeDetector

Commit:
make change detectors chainable, add GradleBuildChangeDetector

Commit:
factor out index filtering from change detectors

gunther-bachmann commented on org.testeditor.web.dropwizard.xtext/src/main/java/org/testeditor/web/xtext/index/changes/ChainableChangeDetector.xtend:
```diff
@@ -0,0 +1,47 @@
+	}
+	
+	private final def ChangedResources handleAndProceed(ResourceSet resourceSet, String[] paths, SetBasedChangedResources detectedChanges) {
+		if (handleChangeDetectionRequest(resourceSet, paths, detectedChanges) && getSuccessor !== null) {
```
I prefer not to have methods actually "doing" something within an if statement
thus: I'd rather write:
val continueDetection = handleChangeDetectionRequest(resourceSet, paths, detectedChanges)
if (continueDetection && getSuccessor !== null) {
...
} else  {
...
}

gunther-bachmann commented on org.testeditor.web.dropwizard.xtext/src/main/java/org/testeditor/web/xtext/index/changes/ChainableChangeDetector.xtend:
```diff
@@ -0,0 +1,47 @@
+ * <i>Template Method</i> pattern, which factors out the chaining / successor
+ * invocation logic.
+ */
+abstract class ChainableChangeDetector implements ChangeDetector {
```
wouldn't it be more practical to have a builder that takes an iterable/list/array of ChangeDetector, and returns a ChangeDetector that has them wrapped in a chain? This would allow to create chains of arbitrary change detectors instead of forcing implementing this abstract class (less hierarchy, more composibility), right?

gunther-bachmann commented on org.testeditor.web.dropwizard.xtext/src/main/java/org/testeditor/web/xtext/index/changes/IndexFilter.xtend:
```diff
@@ -0,0 +1,57 @@
+	override boolean isRelevantForIndex(String filePath) {
+		if (filePath === null) {
+			return false
+		}
```
please use else. we decided to make no use of early returns (after some discussion with Franz ;-)

gunther-bachmann commented on org.testeditor.web.dropwizard.xtext/src/main/java/org/testeditor/web/xtext/index/changes/IndexFilter.xtend:
```diff
@@ -0,0 +1,57 @@
+		}
+		val knownLanguageExtensions = Resource.Factory.Registry.INSTANCE.extensionToFactoryMap.keySet
+		val fileExtension = FilenameUtils.getExtension(filePath).toLowerCase
+		return knownLanguageExtensions.exists[ext|ext.toLowerCase == fileExtension]
```
use compareToIgnoringCase, which carries intend more explicit (no need for toLowerCase)

gunther-bachmann commented on org.testeditor.web.dropwizard.xtext/src/main/java/org/testeditor/web/xtext/index/changes/IndexFilter.xtend:
```diff
@@ -0,0 +1,57 @@
+	var Iterable<Path> searchPaths = null
+
+	override boolean isRelevantForIndex(String path) {
+		return getSearchPaths.exists[new File(path).toPath.toAbsolutePath.startsWith(it)]
```
extract out of loop 'new File(path).toPath.toAbsolutePath' (or into separate method, since it is used similarly below)

gunther-bachmann commented on org.testeditor.web.dropwizard.xtext/src/main/java/org/testeditor/web/xtext/index/persistence/GitService.xtend:
```diff
@@ -23,6 +23,8 @@ import org.slf4j.LoggerFactory
 import static org.eclipse.jgit.lib.ConfigConstants.CONFIG_KEY_URL
 import static org.eclipse.jgit.lib.ConfigConstants.CONFIG_REMOTE_SECTION
 import static org.eclipse.jgit.lib.Constants.*
+import org.eclipse.jgit.treewalk.AbstractTreeIterator
```
organize imports

gunther-bachmann commented on org.testeditor.web.dropwizard.xtext/src/main/java/org/testeditor/web/xtext/index/persistence/GitService.xtend:
```diff
@@ -99,20 +101,51 @@ class GitService {
+	}
+	
+	private def List<DiffEntry> calculateDiff(AbstractTreeIterator oldTree, AbstractTreeIterator newTree) {
+			val diff = git.diff.setOldTree(oldTree).setNewTree(newTree).call
```
indentation

gunther-bachmann commented on org.testeditor.web.dropwizard.xtext/src/test/java/org/testeditor/web/xtext/index/changes/ChainableChangeDetectorTest.xtend:
```diff
@@ -0,0 +1,96 @@
+package org.testeditor.web.xtext.index.changes
```
test is probably not necessary any more, if we choose minimalistic chaining through streaming api

gunther-bachmann commented on org.testeditor.web.dropwizard.xtext/src/test/java/org/testeditor/web/xtext/index/changes/GitBasedChangeDetectorTest.xtend:
```diff
@@ -0,0 +1,141 @@
+		]
+
+		when(mockGit.allFilesAsDiff(headOnInitialUpdate.name())).thenReturn(diffs)
+		when(mockGit.calculateDiff(headOnInitialUpdate.name(), headOnSubsequentUpdate.name())).thenReturn(diffs)
```
this when is used after the first invocation of 'detectChanges', right? please put it after that call below.

Commit:
refactor in response to pull request #30 remarks

Simplified ChangeDetector API by removing the
chain of responsibility pattern and replacing it
with a simple class that depends directly on the
required detectors. While both less generic and
less configurable, it considerably reduces API
complexity, while still retaining a level of
flexibility that should be sufficient for current
requirements.
